src/core/sock.h

```diff
@@ -78,6 +79,9 @@ struct nn_sock
     /*  Next endpoint ID to assign to a new endpoint. */
     int eid;
 
+    /*  Count of active holds against the socket. */
+    int holds;
```

Comment:
It can't really be decomposed.  Really, this is the best solution -- and its a paradigm that has been well tested and battle proven in operating system kernels for decades.